cluster-autoscaler/core/scaledown/legacy/wrapper.go

```diff
@@ -0,0 +1,116 @@
+
+	apiv1 "k8s.io/api/core/v1"
+	policyv1 "k8s.io/api/policy/v1beta1"
+	klog "k8s.io/klog/v2"
+)
+
+// ScaleDownWrapper wraps legacy scaledown logic to satisfy scaledown.Planner &
+// scaledown.Actuator interfaces.
+type ScaleDownWrapper struct {
+	sd        *ScaleDown
+	pdbs      []*policyv1.PodDisruptionBudget
+	processor nodes.ScaleDownNodeProcessor
+}
+
+// NewScaleDownWrapper returns a new ScaleDownWrapper
+func NewScaleDownWrapper(sd *ScaleDown, processor nodes.ScaleDownNodeProcessor) *ScaleDownWrapper {
+	return &ScaleDownWrapper{
+		sd:        sd,
+		processor: processor,
+	}
+}
+
+// UpdateClusterState updates unneeded nodes in the underlying ScaleDown.
+func (p *ScaleDownWrapper) UpdateClusterState(allNodes []*apiv1.Node, deletionsInProgress []string, recentEvictions []*apiv1.Pod, pdbs []*policyv1.PodDisruptionBudget, currentTime time.Time) errors.AutoscalerError {
+	p.sd.CleanUp(currentTime)
+
+	var scaleDownCandidates []*apiv1.Node
+	var podDestinations []*apiv1.Node
+
+	// podDestinations and scaleDownCandidates are initialized based on allNodes variable, which contains only
+	// registered nodes in cluster.
+	// It does not include any upcoming nodes which can be part of clusterSnapshot. As an alternative to using
+	// allNodes here, we could use nodes from clusterSnapshot and explicitly filter out upcoming nodes here but it
+	// is of little (if any) benefit.
+
+	if p.processor == nil {
+		scaleDownCandidates = allNodes
+		podDestinations = allNodes
+	} else {
+		var err errors.AutoscalerError
+		scaleDownCandidates, err = p.processor.GetScaleDownCandidates(p.sd.context, allNodes)
+		if err != nil {
+			klog.Error(err)
+			return err
+		}
+		podDestinations, err = p.processor.GetPodDestinationCandidates(p.sd.context, allNodes)
+		if err != nil {
+			klog.Error(err)
+			return err
+		}
+	}
+	return p.sd.UpdateUnneededNodes(podDestinations, scaleDownCandidates, currentTime, pdbs)
+}
+
+// CleanUpUnneededNodes cleans up unneeded nodes.
+func (p *ScaleDownWrapper) CleanUpUnneededNodes() {
+	p.sd.CleanUpUnneededNodes()
+}
+
+// UnneededNodes returns a list of unneeded nodes.
+func (p *ScaleDownWrapper) UnneededNodes() []*apiv1.Node {
+	return p.sd.UnneededNodes()
+}
+
+// UnremovableNodes returns a list of nodes that cannot be removed.
+func (p *ScaleDownWrapper) UnremovableNodes() []*simulator.UnremovableNode {
+	return p.sd.UnremovableNodes()
+}
+
+// StartDeletion triggers an actual scale down logic.
+func (p *ScaleDownWrapper) StartDeletion(nodes []*apiv1.Node, currentTime time.Time) (*status.ScaleDownStatus, errors.AutoscalerError) {
+	return p.sd.TryToScaleDown(currentTime, p.pdbs)
+}
+
+// DeletionsInProgress returns node names of currently deleted nodes.
+// Current implementation is not aware of the actual nodes names, so it returns
+// a fake node name instead.
+// TODO: Return real node names
+func (p *ScaleDownWrapper) DeletionsInProgress() []string {
+	if p.sd.nodeDeletionTracker.IsNonEmptyNodeDeleteInProgress() {
```

Comment:
I think that eventually a shared implementation of Actuator should exist that takes over the ownership of nodeDeletionTracker.